G0.193 Setup experiment trait picker form

reynoldtan opened the pull request:
**Issue #183 Create Experiment Configuration Page**

## Motivation

<!-- This can usually be copied from the issue. Please do not just say, go see issue but instead copy the relevant details here. -->

This is the second phase of the "Create Experiment Configuration Page" task. This PR introduces the Trait Picker window that enables selection of traits and assign them to experiments.

<img width="741" height="517" alt="Screenshot 2025-11-03 at 8 17 59 AM" src="https://github.com/user-attachments/assets/0e981be6-30e6-4de4-b624-30a4df690d65" />




## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an automated test to ensure it doesn't return.*

1. Setup trait picker window.
2. Add remove handler to remove icon in the trait listing.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

PhenoExperimentTraitPickerFormTest
- Test getFormId() method.
- Test buildForm() method - make sure that fields have the correct render array key - value pairs.
- Test matchTrait() AJAX callback method - test the callback method that will list traits matching a keyword.
- Test page exception scenarios.

PhenoExperimentTraitHandlerControllerTest
- Test AJAX request to assign a trait to experiment.
- Test AJAX request to remove a trait from experiment.
- Test handler exception cases.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

Switch to this branch - **g0.183-ProjectConfigPageAddTrait**

1. Create the following organisms.
Content > Tripal Content > Add Tripal Content > Organism
(**my_site/bio_data/add/organism**)
- Lens culinaris
- Triticum durum


2. Configure the genus (set Trait, Method, Unit and DB) so that it is recognized by the Phenotypes Module.
Tripal > Extensions > TripalCultivate Phenotypes Configuration > Ontology Terms
(**my_site/admin/tripal/extension/tripal-cultivate/phenotypes/ontology**)


3. Create a Research Experiment Tripal Content Type.
Content > Tripal Content > Add Tripal Content > Research Experiment
(**my_site/bio_data/add/research_experiment**)

Provide an experiment name and add the 2 organisms created in step 1 (provide only the genus into the field) into the Germplasm Genus field found in Design tab (use the button Add another item to add a second field). Click Save.

<img width="541" height="477" alt="image" src="https://github.com/user-attachments/assets/27160407-1cae-45c1-a49b-5d0649552f21" />

4. Upload traits. Run Tripal job for each genus.
Tripal > Data Loaders > Tripal Cultivate: Phenotypic Trait Importer
(**my_site/admin/tripal/loaders/trpcultivate-phenotypes-traits-importer**)

[Traits.zip](https://github.com/user-attachments/files/23304608/Traits.zip)

5. Access the Research Experiment.
Content > Tripal Content > The name of the Research Experiment > Phenotypes tab.

<img width="790" height="334" alt="image" src="https://github.com/user-attachments/assets/f12b5380-f0ec-43f1-b693-a208c964f416" />

6. Click <img width="92" height="82" alt="image" src="https://github.com/user-attachments/assets/c2f98388-f23e-4c0b-8491-7a2c002b2619" />

7. If the trait picker is launched without a specified genus from the trait listing page, the window will require to select a genus. Use the search field to search specific trait, or use the "Show All Traits" link to display all available trait for a genus.
<img width="736" height="591" alt="Screenshot 2025-11-03 at 8 18 47 AM" src="https://github.com/user-attachments/assets/1008fb27-bbb9-43b3-939f-767fbba0f336" />

**NOTE**: The "Suggest a Trait" link currently uses **mailto:** implementation. Further discussion is need to determine a better approach.

reynoldtan committed:
Adds event listener to add trait button

reynoldtan committed:
Setup dialog window

reynoldtan committed:
Restores config form

reynoldtan committed:
Restores config form style

reynoldtan committed:
Adjust stacking order of window and overlay

reynoldtan committed:
Improve override code

reynoldtan committed:
Setup search field

reynoldtan committed:
Style trait listings

reynoldtan committed:
Add field controls

reynoldtan committed:
Apply autocomplete cvterm to search field

reynoldtan committed:
Prevent form submission on enter

reynoldtan committed:
Adds event listener when click on a suggestion

reynoldtan committed:
Create show all traits link

reynoldtan committed:
Skip traits that have already been assigned to experiment

reynoldtan committed:
Insert trait @TODO: fix duplicate traits

reynoldtan committed:
Improve code

reynoldtan committed:
Update trait options

reynoldtan committed:
Allow for selection of a genus

reynoldtan committed:
Setup clear form option

reynoldtan committed:
Fully working trait picker. @TODO: delete from summary listing

reynoldtan committed:
Setup remove trait controller

reynoldtan committed:
Remove trait row from listing

reynoldtan committed:
Fix failed test

reynoldtan committed:
Adjust styling

reynoldtan committed:
Final revisions to trait picker

reynoldtan committed:
Setup automated test

reynoldtan committed:
Setup automated test of form and controller

reynoldtan committed:
Setup additional test case

reynoldtan committed:
Fixes failed test

reynoldtan committed:
Fixes failed test - incorrect propery name

reynoldtan committed:
Fixes failed test - undefined this

reynoldtan committed:
Final revisions

reynoldtan committed:
Replaces the word delete with remove

carolyncaron reviewed (commented):
The following is my review from doing a manual test. Overall, I found it very intuitive to use! The functionality seems to work well and the AJAX is smoooooth. 💯 I have just a few comments:
- I feel that the "Add Trait" pop-up is quite small on the screen and not utilizing the space very effectively. I know this is largely based on resolution of the monitor though but it would nice to adapt the size to take up more of the screen if possible. 🙈 
<img width="2554" height="1323" alt="image" src="https://github.com/user-attachments/assets/16670ee5-9823-48f3-af14-a95d32f4d6f6" />

- I LOVE the detail where selecting a genus on the view page before clicking "Add Trait" pre-selects that genus for you when adding traits. I also noticed that once a genus is selected, you can no longer use that dropdown again unless you exit the pop-up window and reopen it. Is this by design or a limitation of AJAX? It would be nice to be able to switch back and forth between genuses if desired (though not crucial to be honest, since more than one genus for a research experiment is expected to be uncommon).
<img width="708" height="162" alt="image" src="https://github.com/user-attachments/assets/05354261-9110-4c36-bcfb-d7b596a78ea0" />

- I'm not sure that the form field for adding a label is clear enough about what its looking for. Perhaps we should include a short description above the traits (right below the search) guiding the user to come up with a good label and give a couple of examples. I also like that it tried to suggest a label for each trait, but in my opinion they are currently too long for what a researcher would find reasonable as column headers in their data. Instead of suggesting both "Trait Name + Method Short Name" I would stick with only Trait Name here.
<img width="630" height="297" alt="image" src="https://github.com/user-attachments/assets/4e30d73f-5a53-4ebb-918c-3ced77c973d3" />

- The tooltip should also mention that the point of the label is to make it the column header of the file - I don't think that currently comes across. How about:
> A short label that will be used as the column header for this Trait-Method-Unit combination in the data collection file for this experiment. This must be unique within this experiment.
 
- Lastly, I agree with finding a better approach for "Suggest a Trait". I clicked on it once and somehow ended up with eight email pop-ups 🤣 
<img width="800" height="903" alt="Screenshot 2025-11-06 at 10 18 18 AM" src="https://github.com/user-attachments/assets/f25ae65c-ec47-4418-a180-b096e7ef1b40" />
I think I pictured a basic form for adding a custom trait (that an admin can approve maybe), but if we want to keep it very simple in the meantime, we could just direct them to the Phenotype Traits Importer?

reynoldtan committed:
Revisions to window size and help text

reynoldtan committed:
Few minor revisions

reynoldtan commented:
Thanks @carolyncaron for the suuuuuuuper thorough review 👍 I made a few revisions to address the window size, help text, and the redirect to trait importer in this commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/c1b8a10c57bab2e286c43c56eb0e9bc39f73e4b2

As for the select genus field being disabled, this is intentional and by design to ensure users remain deliberate and aware of the genus to which they are working with.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+   *
+   * @var array
+   */
+  private array $request;
```
We should be very careful when using private scope! I found private scope caused issues with dependency injection and serialized objects and had to remove it in a number of places in Tripal core. Additionally it can make testing more difficult then needed since private scoped variables will not be accessible in mock objects.

We should only use it if there is a very good reason we don't want any extending classes to touch a variable. The only case I can think of is related to sensitive data and security.

reynoldtan replied:
Noted and this item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+   * @return Symfony\Component\HttpFoundation\JsonResponse
+   *   JSON response array message and status code.
+   */
+  public function handleAction(Request $request, string $action) {
```
I think these docs could use some expanded to make how this is happening more clear. Maybe something like:


```suggestion
   * Handle AJAX request to assign/remove trait to/from experiment.
   *
   * The "trait" and "experiment" this action will apply to is encoded in the
   * request object and will be saved to self::request.
   *
   * @param Symfony\Component\HttpFoundation\Request $request
   *   Request.
   * @param string $action
   *   The requested action to be peformed for a trait. Supported actions are:
   *   - assign: assigns the trait to the experiment.
   *   - remove: removes the trait from the experiment.
   *
   * @return Symfony\Component\HttpFoundation\JsonResponse
   *   JSON response array message and status code.
   */
  public function handleAction(Request $request, string $action) {
```

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+  /**
+   * Request array.
+   *
+   * @var array
```
Expand this docblock to indicate that this is the data from the current request and document the keys that we expect and what they are. For example, 

The following keys are expected:

- project (int): the project_id for the experiment the action should be applied to.

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    $this->request = $request->request->all();
+    $request_params_keys = is_null($this->request) ? [] : array_keys($this->request);
```
Add a comment documenting what these lines are doing so they stand out from the validation checks and so that the developer can follow the logic. Also, add an @see to indicate where they can find documentation on the expected keys.

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+      }
+    }
+
+    $method = $action . 'Trait';
```
Why not make the action = `assignTrait` to begin with? What would make it easier to support other configuration actions in the future that may be related to non-trait parts like the combo label or method/unit.

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+   *   - 'message': the relevant status message.
+   *   - 'status_code': the response status code ie. 200 for Ok.
+   */
+  public function assignTrait() {
```
Document that the trait and experiment being acted on was encoded in the request and is available in self::request. Also indicate which self::request keys specifically this method requires are present.

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Controller/PhenoExperimentTraitHandler.php:
```diff
@@ -0,0 +1,257 @@
+
+  /**
+   * Handle trait callback: remove a trait.
+   *
```
Same as above, please document where the experiment and trait info is and what keys from self::request are required by this method. Especially since the combo_id is required to fit a very specific format!

reynoldtan replied:
This item is no longer relevant as the referenced file is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -280,7 +283,19 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      $remove = [
+        '#type' => 'link',
+        '#title' => '',
+        '#url' => Url::fromUri('internal:#'),
```
Drupal ethos is to never make core functionality javascript only in case someone has it turned off. In this case, that means that we want this URL to point to a route and we want the javascript to use that route.

reynoldtan replied:
This item is no longer relevant as the array is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -318,7 +333,9 @@ public function buildForm(array $form, FormStateInterface $form_state) {
           ],
-          $remove,
+          [
+            'data' => $remove,
```
A comment here about what is in remove at this point would be helpful as I am having trouble following the logic. In the controller I see that the `combo_id` should be in the request data but reading the code here it looks like $remove is either `-` or a full on link render array that contains data-combo_id as an attribute.

reynoldtan replied:
Noted and this item is no longer relevant as the variable is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitPickerForm.php:
```diff
@@ -0,0 +1,452 @@
+   *
+   * @var string
+   */
+  private string $filter_genus;
```
Does this really need to be private?

reynoldtan replied:
This item is no longer relevant as the property is no longer part of the codebase

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitPickerForm.php:
```diff
@@ -0,0 +1,452 @@
+      if (!$research_experiment->id()) {
+        // Research experiment entity does not exist.
+        throw new NotFoundHttpException();
+      }
```
I think this should be outside the if/else shouldn't it? Don't we also want an exception if the id is in the form state and we are unable to load it via the entity type manager? This should also log a watchdog error letting the admin know that someone tried to configure traits for an experiment we were unable to load!

reynoldtan replied:
Watchdog implemented

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,452 @@
+    if (!$experiment) {
+      // Research experiment entity does not exist contain exp_name field.
+      throw new NotFoundHttpException();
+    }
```
This should also log a watchdog error so that the admin knows that we expect there to be an exp_name field!

reynoldtan replied:
Watchdog implemented

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitPickerForm.php:
```diff
@@ -0,0 +1,452 @@
+    }
+
+    $project_id = (int) $experiment[0]['record_id'];
+    $form_state->set('project_id', $project_id);
```
I think this should be saved via a hidden form element (i.e. #type => value) rather then set directly to the form state. That forces it to be available to validation, etc.

reynoldtan replied:
This item is no longer relevant as this variable is no longer part of the codebase.

laceysanderson commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitPickerForm.php:
```diff
@@ -0,0 +1,452 @@
+      'project' => $project_id,
+      'genus' => $genus,
+      'user' => $this->currentUser()->id(),
+    ];
```
This should work via the form submit alone without javascript needing to be triggered.

reynoldtan replied:
This item is no longer relevant as this render array is no longer part of the codebase.

laceysanderson reviewed (dismissed):
Wow this is the most javascript heavy form I have ever seen :see_no_evil: I think it's really important we stick to the form api to process everything and javascript is just used to limit page reloads. Otherwise there is a chance we are bypassing important drupal form api security, etc.

Commit:
Revise remove trait routine

Commit:
Use Drupal AJAX

reynoldtan committed:
use drupal ajax framework

Commit:
revise code

Commit:
Implement filter and remove actions

reynoldtan committed:
Create genus filter jump menu

reynoldtan committed:
Revised forms

reynoldtan committed:
Update config forms

reynoldtan committed:
Final revisions

reynoldtan committed:
Trigger an error on duplicate label

reynoldtan committed:
Manage warning messages

reynoldtan committed:
Replace close button text

reynoldtan committed:
Fix incomplete search result

reynoldtan committed:
Adjust field spacing

reynoldtan committed:
Add field help text

reynoldtan committed:
Inject Tripal logger service

reynoldtan committed:
Field help text revisions

reynoldtan committed:
Query trait revisions

reynoldtan committed:
Trait selector kernel test revisions

reynoldtan committed:
Test load genus traits

reynoldtan committed:
Test add trait combo

reynoldtan committed:
Setup trait selector kernel test

reynoldtan committed:
Setup exceptions test

reynoldtan committed:
Revised trait picker form and kernel test

reynoldtan committed:
Summary table kernel test updates

Commit:
Updated forms

Commit:
Fixes failed test

reynoldtan committed:
Minor text revisions

reynoldtan commented:
This update introduces revisions to Phenotypes Experiment Configuration form suite, with key enhancements focused on these areas.

- Removal of custom CSS and Javascript libraries.
- Unnecessary routing.
- Additional functionality, drop down buttons, and toolbars.
- Use of render array and properties #state and #ajax to manage form behaviours.

and other with the goal of strengthening the overall codebase and ensure it is solid, robust and compliant to Drupal framework.

Below are sources of test data used in the following screenshots:

Organism: Lens _culinaris_ and Triticum _durum_

**Lens Traits:** 
[sample traits-Lens.tsv.zip](https://github.com/user-attachments/files/24483757/sample.traits-Lens.tsv.zip)

**Triticum Traits:**
[sample traits-Triticum.tsv.zip](https://github.com/user-attachments/files/24483759/sample.traits-Triticum.tsv.zip)


**Summary Table**
<img width="3058" height="3954" alt="image" src="https://github.com/user-attachments/assets/4e80690d-cff8-4e84-801e-ebbdcd48fbcd" />

**Trait Selector**
<img width="3088" height="1840" alt="image" src="https://github.com/user-attachments/assets/90354756-12ae-4aa8-8965-132f2b92eab4" />

reynoldtan committed:
Expand field help text

reynoldtan committed:
Make updates happen one time

reynoldtan committed:
Fix a code standard Problem

reynoldtan committed:
Minor revisions

reynoldtan committed:
Merge branch '4.x' into g0.183-ProjectConfigPageAddTrait

reynoldtan committed:
Setup a caching scheme

reynoldtan commented:
Please don't invest significant time reviewing the current summary table, as a more refined version is on the way 

<img width="3058" height="3024" alt="image" src="https://github.com/user-attachments/assets/2272c3ae-bc9a-4006-9ada-587bcaf52038" />


- https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/192

vidu-wil commented:
I did the manual testing for this and it looks great! 👍 All the functionality mentioned in the description works as expected. 
<img width="1921" height="965" alt="Screenshot 2026-01-26 at 2 43 40 PM" src="https://github.com/user-attachments/assets/a43dce71-e442-4cad-b52f-37f47e120cc7" />

✅ I see that the pop-window size issue that Carolyn mentioned above has been fixed and the pop-up window is adapting according to the resolution and window size now.
✅ Also, I like the fact that "Suggest a Trait" is directing to the "Tripal Cultivate: Phenotypic Trait Importer" rather than linking to mailto: option.
❔I just have a minor suggestion about the wording of the "Close & Update Traits" button. I think it makes more sense to have it worded as "Update Traits & Close" or "Update and Close" instead.
<img width="586" height="292" alt="Screenshot 2026-01-26 at 3 38 02 PM" src="https://github.com/user-attachments/assets/cf0c4bfb-4a6d-4d56-9678-824543def019" />

✅ The summary page shows everything as expected and I tried setting some items as required and the drop down next to "Remove" on each trait shows either "Set Require" or "Set Optional" correctly and adapts accordingly after changing.
<img width="1745" height="961" alt="Screenshot 2026-01-26 at 3 02 59 PM" src="https://github.com/user-attachments/assets/e84a2c89-d1f5-4818-83f6-0db397943811" />

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -334,7 +484,123 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+
+    if ($ok === 1) {
+      $this->messenger()
+        ->addStatus('The trait combo operation ' . ucfirst($action) . ' completed successfully.');
```
```suggestion
        ->addStatus("The trait combo operation '" . ucfirst($action) . "' completed successfully.");
```

reynoldtan replied:
Addressed in commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/a93bf15ccacb3678842234c22f1c8172dcf43a6e. I flipped the quotes so that the text to be highlighted is in double quotes instead of single, as it is familiar to user to denote important string

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+   *   TripalCultivate Phenotypes Genus-Ontology.
+   * @param \Drupal\trpcultivate_phenotypes\Service\TripalCultivatePhenotypesGenusProjectService $service_PhenoGenusProject
+   *   TripalCultivate Phenotypes Genus-Project service.
+   * @param \Drupal\trpcultivate_phenotypes\Service\ripalCultivatePhenotypesTraitsService $service_PhenoTraits
```
```suggestion
   * @param \Drupal\trpcultivate_phenotypes\Service\TripalCultivatePhenotypesTraitsService $service_PhenoTraits
```

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+    $experiment = $tripal_entity->get('exp_name')->getValue();
+    ['record_id' => $experiment_id, 'value' => $experiment_name] = $experiment[0];
+
+    // Update the title to show which reseach experiment is being configured.
```
```suggestion
    // Update the title to show which research experiment is being configured.
```

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+        'a_tip' => [
+          '#type' => 'html_tag',
+          '#tag' => 'p',
+          '#value' => 'Start typing part of the trait name into the search field to search for specific traits, or click - Show all Traits button, to explore all available traits for the selected genus.',
```
```suggestion
          '#value' => "Start typing part of the trait name into the search field to search for specific traits, or click 'Show all Traits' button, to explore all available traits for the selected genus.",
```

reynoldtan replied:
Addressed in commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/a93bf15ccacb3678842234c22f1c8172dcf43a6e. I flipped the quotes so that the text to be highlighted is in double quotes instead of single, as it is familiar to user to denote important string

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+    // Capture trigger element responsible for staring trait selection request.
+    $trigger_el = $form_state->getTriggeringElement() ?? 0;
+
+    // Prepare tratis that matched the search key.
```
```suggestion
    // Prepare traits that matched the search key.
```

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+          $form[$form_dialog_wrapper]['duplicate_label'] = [
+            '#theme' => 'status_messages',
+            '#message_list' => [
+              'warning' => [
```
```suggestion
              'error' => [
```
I believe this should be an error instead of warning. (The '#status_headings' below is defined as an 'error' and the heading does not appear due to this)
<img width="645" height="920" alt="Screenshot 2026-01-27 at 11 39 14 AM" src="https://github.com/user-attachments/assets/23dbd8ee-ff22-4416-a398-a3c3b10a4b86" />

reynoldtan replied:
great catch, here's an example
<img width="1664" height="996" alt="image" src="https://github.com/user-attachments/assets/eaf4ce3d-a0e9-4c91-9c11-c08c12a056ba" />

vidu-wil commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -272,60 +386,96 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      $form[$summary_table_name][$i]['trait_combo'] = [
+        '#type' => 'component',
+        '#component' => 'trpcultivate_phenotypes:trait_combo',
+        '#slots' => [],
```
```suggestion
```
'slots' has been removed from trait_combo.component.yml

reynoldtan replied:
good catch 👍

vidu-wil commented on trpcultivate_phenotypes/tests/src/Kernel/Forms/PhenoExperimentTraitSelectorFormTest.php:
```diff
@@ -0,0 +1,894 @@
+  }
+
+  /**
+   * Test valid reoute parameter and page exceptions.
```
```suggestion
   * Test valid route parameter and page exceptions.
```

vidu-wil commented on trpcultivate_phenotypes/tests/src/Kernel/Forms/PhenoExperimentConfigurationFormTest.php:
```diff
@@ -736,114 +885,54 @@ public function testPageAccess(string $scenario, array $user, array $expected) {
 
   /**
-   * Test page exceptions.
+   * Test valid reoute parameter and page exceptions.
```
```suggestion
   * Test valid route parameter and page exceptions.
```

reynoldtan committed:
Update trpcultivate_phenotypes/tests/src/Kernel/Forms/PhenoExperimentConfigurationFormTest.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/tests/src/Kernel/Forms/PhenoExperimentTraitSelectorFormTest.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php

Co-authored-by: Vidusha Wilpita <132868285+vidu-wil@users.noreply.github.com>

reynoldtan committed:
text revisions

reynoldtan committed:
Merge branch 'g0.183-ProjectConfigPageAddTrait' of https://github.com/TripalCultivate/TripalCultivate-Phenotypes into g0.183-ProjectConfigPageAddTrait

reynoldtan commented:
> ❔I just have a minor suggestion about the wording of the "Close & Update Traits" button. I think it makes more sense to have it worded as "Update Traits & Close" or "Update and Close" instead.

I set the label to "Close and Update Traits" to reflect the actual sequence of events triggered when the button is clicked - window closes first, then as second event the summary table reloads (update traits). I initially tried the alternative "Update & Close", but found it ambiguous, while still on the selector form, I kept questioning what was being updated. It is good you highlighted this as it can now serve a focus point for other reviewers to ensure interface is clear.

reynoldtan committed:
Fixes failed test

ruobinLiu commented:
Everything worked as expected! Nice work Reynold!

Except I observed some minor differences during my test than the screenshots, have them both circled in red:

**Step 5**
Seems like I can add trait by clicking on the "Add Trait" botton
<img width="896" height="445" alt="543112069-715fc2a8-38a4-4025-9bd6-192da75bc338" src="https://github.com/user-attachments/assets/b8ede663-d6ae-423e-9af0-73f487603d36" />

**Step 7**
There is a toggle that could allow me to disable a trait?
<img width="902" height="808" alt="543112726-4f0cdbdb-9c6b-4479-b36a-e6b617fcb0cf" src="https://github.com/user-attachments/assets/e24b4712-6c7f-45e3-baaf-e3a3ab3976bf" />

reynoldtan commented:
Thanks for the review @ruobinLiu 👍 

> Everything worked as expected! Nice work Reynold!
> 
> Except I observed some minor differences during my test than the screenshots, have them both circled in red:
> 
> **Step 5** Seems like I can add trait by clicking on the "Add Trait" botton <img alt="543112069-715fc2a8-38a4-4025-9bd6-192da75bc338" width="896" height="445" src="https://private-user-images.githubusercontent.com/45402954/543114315-b8ede663-d6ae-423e-9af0-73f487603d36.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAwNDk4NTAsIm5iZiI6MTc3MDA0OTU1MCwicGF0aCI6Ii80NTQwMjk1NC81NDMxMTQzMTUtYjhlZGU2NjMtZDZhZS00MjNlLTlhZjAtNzNmNDg3NjAzZDM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAyVDE2MjU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMmFjN2FkOTUxZTc0YjA1Yzc2NDMyMDFjNjRjZWQzMGNmNjZjNGU1ZWYwYmNmNWJhOTJkODY5NDBkOWNlNTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5nY9_3RgS-sDJEu9r3Ds9cOKBGXHI5U9Z5gHxhy-3QY">
> 

I added a quick link, with the same action as the main + Add Trait button, if there is not trait combo in the summary table as an alternative way to add trait.

> **Step 7** There is a toggle that could allow me to disable a trait? <img alt="543112726-4f0cdbdb-9c6b-4479-b36a-e6b617fcb0cf" width="902" height="808" src="https://private-user-images.githubusercontent.com/45402954/543114337-e24b4712-6c7f-45e3-baaf-e3a3ab3976bf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAwNDk4NTAsIm5iZiI6MTc3MDA0OTU1MCwicGF0aCI6Ii80NTQwMjk1NC81NDMxMTQzMzctZTI0YjQ3MTItNmM3Zi00NWUzLWJhYWYtZTNhM2FiMzk3NmJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAyVDE2MjU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZGVjNTVkMDQ5YzVmYzBjYjEyYTQxZjc0ZTI0MmZlY2ZhY2YwNTQ4Y2VkNDdlZTQ4YmFiMzY2OGFhNTUwZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.I8Ryz-CDoM0JaBNaJ9yr0iZevbz-5liB0wESJE1u4z4">

The toggle/checkbox field is to indicate that the combo your adding to the experiment, is going to be a required trait (it must exists in the data file).

Tip: if you hover on the star ⭐  icon, it will show a brief description.
<img width="581" height="151" alt="Screen Shot 2026-02-02 at 10 36 37 AM" src="https://github.com/user-attachments/assets/ec3531eb-e461-4b39-a328-cf1d0d388b99" />

Adding the trait with the toggle on/checked will attach the same star ⭐  icon in the summary table.
<img width="802" height="160" alt="Screen Shot 2026-02-02 at 10 39 18 AM" src="https://github.com/user-attachments/assets/fa91d693-b574-4ca0-8e09-4187cc6b5b66" />

carolyncaron commented:
> > ❔I just have a minor suggestion about the wording of the "Close & Update Traits" button. I think it makes more sense to have it worded as "Update Traits & Close" or "Update and Close" instead.
> 
> I set the label to "Close and Update Traits" to reflect the actual sequence of events triggered when the button is clicked - window closes first, then as second event the summary table reloads (update traits). I initially tried the alternative "Update & Close", but found it ambiguous, while still on the selector form, I kept questioning what was being updated. It is good you highlighted this as it can now serve a focus point for other reviewers to ensure interface is clear.

How about "Close and Refresh Table"? I think the "Update" word is confusing because it makes it sound like it's "Saving" the traits, but in practice, they are already saved when the "Add" button is clicked. Plus, "Refresh Table" is already used on the summary page, so this would be keep it consistent.

carolyncaron commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -2,29 +2,47 @@
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected Connection $database_connection;
```
I recommend calling this `$drupal_connection` instead to keep consistency with Tripal and make it explicit what type of database connection this is.

reynoldtan replied:
Addressed in commit - https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/112810af3fb7ba4c11f5feb5d2e2a5d48c041bc5

carolyncaron commented on trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php:
```diff
@@ -2,29 +2,47 @@
+   *
+   * @var \Drupal\tripal\Services\TripalLogger
+   */
+  protected $tripal_logger;
```
```suggestion
  protected TripalLogger $tripal_logger;
```

carolyncaron commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected Connection $database_connection;
```
```suggestion
  protected Connection $drupal_connection;
```

reynoldtan replied:
addressed in commit: https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/112810af3fb7ba4c11f5feb5d2e2a5d48c041bc5

carolyncaron commented on trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php:
```diff
@@ -0,0 +1,670 @@
+   *
+   * @var \Drupal\tripal\Services\TripalLogger
+   */
+  protected $tripal_logger;
```
```suggestion
  protected TripalLogger $tripal_logger;
```

carolyncaron reviewed (changes requested):
Thanks for implementing these changes! I did a quick code review, and here are some things that I noticed during the manual review:

- For my test, I added some traits for Lens but none for Triticum yet. If I filter the summary table for Triticum, it says "No traits found for this research experiment" which isn't true because there are some for Lens. Is it possible to mention the genus being filtered in the error message? Something like "No traits for Triticum found for this research experiment. I acknowledge this is a bit tricky because there is a filter for "All Genus", so the current error message works for that case specifically.

- I think it makes more sense now to rename "Suggest a Trait" to be "Import Traits" instead since it is being redirected to the Traits Importer. In the future, we may want to add the ability to "Suggest a trait" using a form, but that's outside the scope of this PR so I think it's best to rename it to accurately reflect what it does.

- When closing the trait picker form, is it possible to make the "X" (Close) option in the top right of the window perform the same action as "Close & Update Traits" currently does? Right now, it just closes the form and the user might wonder what happened to the traits they just added. 
   <img width="1055" height="360" alt="Image" src="https://github.com/user-attachments/assets/8fe5b440-6d08-4da0-a11b-ffba15fd1dc6" />

- The suggested default for Label is still WAY too long in my opinion. I would just go with the Trait Name since that it most realistic in practice. Perhaps @ruobinLiu could confirm here, but the "Method Short Name" in most cases won't be as brief as what your example files show (ex. PLHT). Regardless, combining the Trait Name and Method name still makes this too long in my opinion.

- This is a very minor visual critique, but it looks odd that the 2 options next to the "Remove" button don't have the same size background when the mouse is overlaid on them. This also occurs when the Trait is already required and the text says "Set Optional".
Set Require: <img width="269" height="172" alt="Image" src="https://github.com/user-attachments/assets/e11d6c07-a9be-47e1-b583-ecb3529e5eb8" />
Set Archive: <img width="233" height="164" alt="Image" src="https://github.com/user-attachments/assets/b5118474-e079-4ec7-95db-d83e4f1fe68d" />

- Related to above, I would change "Set Require", "Set Optional" and "Set Archive" to be "Set as required", "Set as optional", and "Archive Trait", respectively.

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentTraitSelectorForm.php

Co-authored-by: Carolyn T Caron <carolyn.caron@usask.ca>

reynoldtan committed:
Update trpcultivate_phenotypes/src/Form/PhenoExperimentConfigurationForm.php

Co-authored-by: Carolyn T Caron <carolyn.caron@usask.ca>

Commit:
Fix misleading no trait error message

reynoldtan commented:
> * For my test, I added some traits for Lens but none for Triticum yet. If I filter the summary table for Triticum, it says "No traits found for this research experiment" which isn't true because there are some for Lens. Is it possible to mention the genus being filtered in the error message? Something like "No traits for Triticum found for this research experiment. I acknowledge this is a bit tricky because there is a filter for "All Genus", so the current error message works for that case specifically.

Addressed in https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/19ec6e3e2e305d6dae5d697981bf1944d4a03bab

Some examples:
<img width="708" height="162" alt="Screen Shot 2026-02-03 at 4 46 19 PM" src="https://github.com/user-attachments/assets/d9faec07-d63d-4391-9ab9-78da3ebe08cc" />
<img width="687" height="159" alt="Screen Shot 2026-02-03 at 4 46 47 PM" src="https://github.com/user-attachments/assets/cce10fc6-72e1-4957-b771-9e17c8fc7bd2" />

Commit:
Rename suggest trait link

Commit:
Rename link title

reynoldtan commented:
> * I think it makes more sense now to rename "Suggest a Trait" to be "Import Traits" instead since it is being redirected to the Traits Importer. In the future, we may want to add the ability to "Suggest a trait" using a form, but that's outside the scope of this PR so I think it's best to rename it to accurately reflect what it does.

Addressed in commit: https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/32555f1f9f80de1169e8079387a7dd5771786834

<img width="817" height="76" alt="Screen Shot 2026-02-03 at 4 50 33 PM" src="https://github.com/user-attachments/assets/7cb0f9ee-16bb-4f1e-90cf-467267074945" />

reynoldtan committed:
Add event listener to dialog close

reynoldtan commented:
> * When closing the trait picker form, is it possible to make the "X" (Close) option in the top right of the window perform the same action as "Close & Update Traits" currently does? Right now, it just closes the form and the user might wonder what happened to the traits they just added.

Addressed in commit: https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/7282d6676de075aabde3d2bfa922086c04da795d

reynoldtan commented:
> * The suggested default for Label is still WAY too long in my opinion. I would just go with the Trait Name since that it most realistic in practice. Perhaps @ruobinLiu could confirm here, but the "Method Short Name" in most cases won't be as brief as what your example files show (ex. PLHT). Regardless, combining the Trait Name and Method name still makes this too long in my opinion.


I see your point on the label length. The reason we combine Trait Name and Method Short Name is to ensure uniqueness. With just the Trait Name, different methods of Plant Height would end up with identical labels - 'Plant Height', and the interface would catch this.

reynoldtan commented:
> * This is a very minor visual critique, but it looks odd that the 2 options next to the "Remove" button don't have the same size background when the mouse is overlaid on them. This also occurs when the Trait is already required and the text says "Set Optional".

Not sure how to rectify this but I looks like this is admin theme issue also in content type view.
<img width="648" height="553" alt="Screenshot 2026-02-04 at 8 01 14 AM" src="https://github.com/user-attachments/assets/59794fd1-d8f1-43df-8eaa-08a25802f405" />

<img width="665" height="649" alt="Screenshot 2026-02-04 at 8 00 32 AM" src="https://github.com/user-attachments/assets/11ff0233-81f2-4538-8e24-49cb4dd289ad" />

reynoldtan committed:
Rename operations

reynoldtan commented:
> * Related to above, I would change "Set Require", "Set Optional" and "Set Archive" to be "Set as required", "Set as optional", and "Archive Trait", respectively.

Addressed in commit : https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/d6cc18797eb03c8571a34e750314aed02981b17a

reynoldtan committed:
Rename database connection property name

reynoldtan committed:
Rename close button

ruobinLiu commented:
> Thanks for the review @ruobinLiu 👍
> 
> > Everything worked as expected! Nice work Reynold!
> > Except I observed some minor differences during my test than the screenshots, have them both circled in red:
> > **Step 5** Seems like I can add trait by clicking on the "Add Trait" botton <img alt="543112069-715fc2a8-38a4-4025-9bd6-192da75bc338" width="896" height="445" src="https://private-user-images.githubusercontent.com/45402954/543114315-b8ede663-d6ae-423e-9af0-73f487603d36.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAwNDk4NTAsIm5iZiI6MTc3MDA0OTU1MCwicGF0aCI6Ii80NTQwMjk1NC81NDMxMTQzMTUtYjhlZGU2NjMtZDZhZS00MjNlLTlhZjAtNzNmNDg3NjAzZDM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAyVDE2MjU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMmFjN2FkOTUxZTc0YjA1Yzc2NDMyMDFjNjRjZWQzMGNmNjZjNGU1ZWYwYmNmNWJhOTJkODY5NDBkOWNlNTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5nY9_3RgS-sDJEu9r3Ds9cOKBGXHI5U9Z5gHxhy-3QY">
> 
> I added a quick link, with the same action as the main + Add Trait button, if there is not trait combo in the summary table as an alternative way to add trait.
> 
> > **Step 7** There is a toggle that could allow me to disable a trait? <img alt="543112726-4f0cdbdb-9c6b-4479-b36a-e6b617fcb0cf" width="902" height="808" src="https://private-user-images.githubusercontent.com/45402954/543114337-e24b4712-6c7f-45e3-baaf-e3a3ab3976bf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAwNDk4NTAsIm5iZiI6MTc3MDA0OTU1MCwicGF0aCI6Ii80NTQwMjk1NC81NDMxMTQzMzctZTI0YjQ3MTItNmM3Zi00NWUzLWJhYWYtZTNhM2FiMzk3NmJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAyVDE2MjU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZGVjNTVkMDQ5YzVmYzBjYjEyYTQxZjc0ZTI0MmZlY2ZhY2YwNTQ4Y2VkNDdlZTQ4YmFiMzY2OGFhNTUwZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.I8Ryz-CDoM0JaBNaJ9yr0iZevbz-5liB0wESJE1u4z4">
> 
> The toggle/checkbox field is to indicate that the combo your adding to the experiment, is going to be a required trait (it must exists in the data file).
> 
> Tip: if you hover on the star ⭐ icon, it will show a brief description. <img alt="Screen Shot 2026-02-02 at 10 36 37 AM" width="581" height="151" src="https://private-user-images.githubusercontent.com/15472253/543896941-ec3531eb-e461-4b39-a328-cf1d0d388b99.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyMzQ1NDksIm5iZiI6MTc3MDIzNDI0OSwicGF0aCI6Ii8xNTQ3MjI1My81NDM4OTY5NDEtZWMzNTMxZWItZTQ2MS00YjM5LWEzMjgtY2YxZDBkMzg4Yjk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA0VDE5NDQwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4NDU5NzM1MTQyY2ViNGZiNGQ2NGVmZDhmNmEzYWY1ODJiNmM2MzllMDc2NzAzMmFkNGM3YmFhMGE4ZjkyNmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.W9UcEnadBULNwvudfSCAnW61B3wTMIWDPn_CeLWx11c">
> 
> Adding the trait with the toggle on/checked will attach the same star ⭐ icon in the summary table. <img alt="Screen Shot 2026-02-02 at 10 39 18 AM" width="802" height="160" src="https://private-user-images.githubusercontent.com/15472253/543897894-fa91d693-b574-4ca0-8e09-4187cc6b5b66.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAyMzQ1NDksIm5iZiI6MTc3MDIzNDI0OSwicGF0aCI6Ii8xNTQ3MjI1My81NDM4OTc4OTQtZmE5MWQ2OTMtYjU3NC00Y2EwLThlMDktNDE4N2NjNmI1YjY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA0VDE5NDQwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkZjBlOGEzYzE0NWQ5MDg5NzgwM2E4MjAxMDYzOWYzNmVhNWE0NzcwMzYwZTFjNzIxNGM2MGQ1ZDgyODhkOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OsuXyWS7li7ZRTqD8NBMKA42nqPQJRmFoPmxnvTwmCM">

Really like the ⭐ icon allowing customized "Required Traits" 👍

ruobinLiu commented:
Step 2 Configure the genus
<img width="714" height="281" alt="Screenshot 2026-02-06 at 3 45 19 PM" src="https://github.com/user-attachments/assets/01acbc38-5143-421f-acb5-43cc86ee1b3e" />

I notice the Trait Picker allows new traits to be added. Do I get to describe this new trait with Pgenotypic Trait File later on?

reynoldtan commented:
> I notice the Trait Picker allows new traits to be added. Do I get to describe this new trait with Pgenotypic Trait File later on?

This option is for traits not defined (or added through the trait picker) and detected in the file upload process, much like the new traits detected in rawphenotypes which required user to describe. This option is not fully implemented in our new phenotypes importer.

carolyncaron commented on trpcultivate_phenotypes/tests/src/Kernel/Forms/PhenoExperimentConfigurationFormTest.php:
```diff
@@ -73,16 +81,11 @@ class PhenoExperimentConfigurationFormTest extends ChadoTestKernelBase {
 
   /**
-   * Expected research experiment entity id to describe a setup.
+   * The table name.
```
Which table? Is this the name of custom table for storing phenotype trait/method/unit combinations?

reynoldtan replied:
expanded the table name in commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/7ac157d8cf99d56c9e808d00430dbb16ec4f1b31

carolyncaron reviewed (changes requested):
Thank you for those changes, the form looks like its just about there!

Just a couple more suggestions:
- Next to "Import Traits", can you add the icon that indicates the link will open in a new tab? I think it's a nice touch when the user knows to expect a new tab will be opened. This is the class I use on RE pages when I link out to JBrowse: `<i class="fa-solid fa-arrow-up-right-from-square"></i>`
- For the status messages when traits are set as required, optional, archived, and restored, I think the message can be more simple and exclude "The trait combo operation". We should avoid using the word "combo" when it is user-facing. 

For example:
<img width="677" height="123" alt="Image" src="https://github.com/user-attachments/assets/dd5ee51c-f13d-4c17-bd26-8d1d60aaa132" />
This message could read as `"Set as Required" completed successfully.`

reynoldtan committed:
Rephrase action message

reynoldtan commented:
> * For the status messages when traits are set as required, optional, archived, and restored, I think the message can be more simple and exclude "The trait combo operation". We should avoid using the word "combo" when it is user-facing.

Addressed in commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/7ac157d8cf99d56c9e808d00430dbb16ec4f1b31

reynoldtan committed:
use new window icon

reynoldtan commented:
> * Next to "Import Traits", can you add the icon that indicates the link will open in a new tab? I think it's a nice touch when the user knows to expect a new tab will be opened. This is the class I use on RE pages when I link out to JBrowse: `<i class="fa-solid fa-arrow-up-right-from-square"></i>`

<img width="1786" height="1050" alt="image" src="https://github.com/user-attachments/assets/5bfd2825-7b1f-4ae3-ba16-b359750ec625" />

Addressed in commit - https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/193/commits/f2ed04b4c1efd5117b5cc54dbcad239afdc195b9

carolyncaron reviewed (approved):
Thanks for those changes! PR approved ✅

reynoldtan commented:
I have reviewed Lacey's past comments and either it is no longer relevant as the file referenced is no longer part of the codebase or I have addressed it fully. Ready to merge.
